src/spline2/MultiBsplineData.hpp

```diff
@@ -31,9 +31,24 @@ namespace spline2
 template<typename T, typename TRESIDUAL>
 inline void getSplineBound(T x, TRESIDUAL& dx, int& ind, int ng)
 {
-  T ipart;
-  dx=std::modf(x,&ipart);
-  ind = std::min(std::max(int(0),static_cast<int>(ipart)),ng);
+  // lower bound
+  if (x < 0)
+  {
+    ind = 0;
+    dx = T(0);
+  }
+  else
+  {
+    T ipart;
+    dx=std::modf(x,&ipart);
+    ind = static_cast<int>(ipart);
+    // upper bound
+    if (ind >= ng)
+    {
+      ind = ng;
+      dx = T(1) - std::numeric_limits<T>::epsilon();
```

Comment:
This is the behavior I think is strange.  Imagine ng=2 I pass x=2.5. the periodic function is sin(x/pi) but I get back 0. Now with the > there I get the value of at 0.5 but at 3.0 I suddenly get the value of 2.0.

Reply:
ng is not the grid point. It is the upper bound. So > is correct, >= is not correct.

if there is n grid point. If in the 3D periodic spline case, 
getSplineBound(,,,ng=n-1)
in the 1D open BC spline case,
getSplineBound(,,,ng=n-2)

in your example, ng=2, x=2.5. [0,3) is the allowed range.
you get integer part 2 and fractional part 0.5
in your example, ng=2, x=3.0. [0,3) is the allowed range.
you get integer part 2 and fractional part 1-epsilon.

Does this make sense?

Reply:
Yes that makes sense but it must be documented is some clear way upper bound is not clear ng is the highest grid point of the implementation.  This seems like its a private static member function of some type class you're avoiding defining. The code could be more self documenting right now its a collection of functions from einspline.  Are we going somewhere like that?

Reply:
We did enough with wrapping einspline with C++ in the past and spline2 is the templated C++ version and we are moving away from einspline. The getSplineBound is used by 1D and 3D cases and I made it a free template function which can be always inlined.